### PR TITLE
Add Customizable MetricsReporters to the Measure-Builds Gradle Plugin

### PR DESCRIPTION
**Pull request by takahirom**
Thank you for the great plugin! I find this plugin to be well-designed and modern, making appropriate use of the latest Gradle features. It's a great example of a plugin that leverages Gradle's capabilities effectively.

# Overview

This pull request introduces a new feature that allows users to customize the `MetricsReporters` in the Measure-Builds Gradle plugin. With this change, users can now provide their own implementation of the MetricsReporter interface to handle the reporting of build metrics according to their specific requirements.

# Note

Please note that with the removal of the built-in reporting functionality, users will now need to provide their own implementation of `MetricsReporter` to handle the reporting of build metrics. This change gives users complete control over how and where the metrics are reported.

I understand that I can fork this plugin and maintain my own version with the changes I need. However, I believe that making the plugin customizable and extensible is beneficial for the open-source community. By working together, we can create a better tool that caters to a wider range of use cases and requirements.

### CODE REVIEW

**Comment by takahirom** [comment]
@wzieba @malinajirka 
Could you take the time to review this PR?

File: measure-builds/src/test/java/com/automattic/android/measure/BuildTimePluginTest.kt [hunk 1]
```diff
@@ -11,37 +11,6 @@ import java.io.File
...
-        ).contains("Calculating task graph as no configuration cache is available for tasks")
-            .contains("Configuration cache entry stored")
-
-        // when
-        val buildUsingConfigurationCache =
-            runner.withArguments("--configuration-cache", "help", "--debug").build()
-
-        // then
-        assertThat(buildUsingConfigurationCache.output).contains("Reusing configuration cache")
-            .contains("Reporting build data to Apps Metrics...")
```
Comment (takahirom) [inline measure-builds/src/test/java/com/automattic/android/measure/BuildTimePluginTest.kt:40]: The new reporter doesn't have the "Reporting build data to Apps Metrics..." message, so I removed this test.

**Comment by wzieba** [comment]
hi @takahirom ! Thank you for the contribution. I agree with your arguments about making this plugin customizable. Initially, we thought it'll be for internal usage only, but that's true it could be nicely extended by introducing custom reporters. 

In the following days, I'll review proposed changes 👍

**Comment by takahirom** [comment]
Hi @wzieba 
Thank you for considering the introduction of custom reporters!
I have worked on the refactoring and discovered a bug where the users' reports were not properly handled through configuration changes.
I believe the issue is resolved now. I'm referring to the buildScan's buildScanPublished{} extension. However, we're unable to view the buildScan's implementation. It seems that the Action interface is utilized within this extension, which is why I opted for Property<Action>.

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt [hunk 2]
```diff
@@ -22,38 +20,38 @@ import io.ktor.http.HttpHeaders
...
-class MetricsReporter(
-    private val logger: Logger,
-    private val authToken: Provider<String>,
-    private val buildDir: File,
-) {
-    suspend fun report(
-        report: InMemoryReport,
-        gradleScanId: String?
+
+object InternalA8cCiReporter {
```
Comment (wzieba) [inline measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt:31]: In the future, we can think of extracting this to a separate plugin, to not ship a core plugin with A8c-specific reporter.

File: measure-builds/src/test/java/com/automattic/android/measure/BuildTimePluginTest.kt (hunk 1 above)
Comment (wzieba) [reply measure-builds/src/test/java/com/automattic/android/measure/BuildTimePluginTest.kt]: Thanks for brining back A8c reporter in form of `InternalA8cCiReporter` and reverting removal. I'll mark this comment as resolved.

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt [hunk 3]
```diff
@@ -22,38 +20,38 @@ import io.ktor.http.HttpHeaders
...
+            report(metricsReport, projectName, authToken)
+        }
+    }
 
-        val payload = report.toAppsMetricsPayload(gradleScanId)
+    suspend fun report(
+        metricsReport: MetricsReport,
+        projectName: String,
+        authToken: String?,
+    ) {
```
Comment (wzieba) [inline measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt:49]: It is tempting to make some `Report` interface and pass reporter-specific parameters (like `projectName` or `authToken` here) as constructor parameters, as you did in 3bb9601c12c699d47f36036c88bfc60a1faa20e4 with `MetricsReporter`.

I like the current approach with `buildMetricsPreparedAction` more, though, as it's not unnecessarily constraining the implementation. Great job 👍 !

**Review by wzieba — COMMENTED** [review]
Thank you for your solid work @takahirom on this feature! I'm after the first round of review and I haven't spotted any blocking request changes.

I'll experiment with the implementation more on our repositories to make sure lack of regression. In the meantime, I've run CI checks and some code styling checks are failing. You can fix more of them by running ` gw :measure-builds:detekt --auto-correct`

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt [hunk 4]
```diff
@@ -22,38 +20,38 @@ import io.ktor.http.HttpHeaders
...
-    private val authToken: Provider<String>,
-    private val buildDir: File,
-) {
-    suspend fun report(
-        report: InMemoryReport,
-        gradleScanId: String?
+
+object InternalA8cCiReporter {
+    private val logger =
+        org.gradle.api.logging.Logging.getLogger(InternalA8cCiReporter::class.java)
```
Comment (wzieba) [inline measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt:33]: Minor nitpick: can we extract import so we'll have `Logging.getLogger(...)` here only?

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt (hunk 3 above)
Comment (takahirom) [reply measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt:49]: Thank you. Implementing a configuration cache can be challenging when using instances, and errors are likely to occur.

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt (hunk 4 above)
Comment (takahirom) [reply measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt]: We already have Ktor Logging import so I used the full package name but we can use an import alias. So I added the type alias here https://github.com/Automattic/measure-builds-gradle-plugin/pull/38/commits/1accd072d3d243f9fab57033780d2ab0d53d3692

**Comment by takahirom** [comment]
Hi @wzieba, I've utilized the detekt autocollect task and added a typealias. If you have any feedback or suggestions, please let me know. Feel free to modify my implementation by making changes directly in this PR.

File: measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt (hunk 4 above)
Comment (wzieba) [reply measure-builds/src/main/java/com/automattic/android/measure/reporters/InternalWooCommerceA8cCiReporter.kt]: Oh right, sorry I didn't check it before suggesting the change. The alias looks nice though, thanks 👍

**Comment by wzieba** [comment]
hi @takahirom , thank you for applying the suggestions. I really like this new API and the flexibility that it brings 🙌 .

While experimenting with the new API introduced in this PR though, I've noticed a problem when using it with Groovy. I've prepared an explanation and fix in #43. Please have no hesitation to leave comments and/or suggestions there.

And just a heads-up - I need to perform a few more tests before merging this PR, but I'll be able to do this only early next week. I hope that's ok for you. In any case, if you'd like, you can check a snapshot of the plugin built from the mirror branch (#39) using coordinates like:

```groovy
plugins {
    id 'com.automattic.android.measure-builds' version '39-1accd072d3d243f9fab57033780d2ab0d53d3692'
}
```

In the future, I plan to make this plugin available from Gradle Plugin Portal (#42) so it'll be easier.

**Review by wzieba — APPROVED** [review]

**Comment by wzieba** [comment]
Thanks again! Looks good with #43 and #44 fixes.

**Pull request merged**